File: 55. Jump Game.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        if (nums.size() < 2)
            return true;
        int i = 0, pos = 0, new_pos = 0;
        while (true) {
            new_pos = pos;
            while (i <= pos) {
                new_pos = max(new_pos, i+nums[i]);
                i++;
            }
            if (new_pos == pos) return false;
            pos = new_pos;
            if (pos >= nums.size() - 1)
                return true;
        }
        
        return false;
    }
};

File: python/55. Jump Game.py
# -*- coding: utf-8 -*-

class Solution(object):
    def canJump(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        max_dis = 0 # 记录下标为i时，到达的最大距离
        for i in range(len(nums)):
            if i > max_dis:
                return False
            max_dis = max(max_dis, i + nums[i])
            if max_dis >= len(nums)-1:
                return True
        return True

# 思路：贪心